File: backend/util/quickwit.py
import psycopg
import requests
import json
import time
from typing import List, Union


from constants import QUICKWIT_ENDPOINT
from models.util import postgres_connection_string

import logging

logger = logging.get_Logger(__name__)

db_string = postgres_connection_string
def print_response(r):
    logger.info(f"quickwit_api_call:\nstatus:{r.status_code}\nresponse:\n{r.text}")


def create_dockets_quickwit_index() -> None:
    request_data = {
        "version": "0.7",
        "index_id": "dockets",
        "doc_mapping": {
            "mode": "dynamic",
            "dynamic_mapping": {
                "indexed": True,
                "stored": True,
                "tokenizer": "default",
                "record": "basic",
                "expand_dots": True,
                "fast": True
            },
            "field_mappings": [
                {
                    "name": "text",
                    "type": "text",
                    "fast": True
                },
                {
                    "name": "timestamp",
                    "type": "datetime",
                    "input_formats": ["unix_timestamp"],
                    "fast_precision": "seconds",
                    "fast": True
                },
            ],
            "timestamp_field": "timestamp"
        },

        "search_settings": {
            "default_search_fields": [
                "text", "state", "city", "country"],
        },
        "indexing_settings": {
            "merge_policy": {
                "type": "limit_merge",
                "max_merge_ops": 3,
                "merge_factor": 10,
                "max_merge_factor": 12
            },
            "resources": {
                "max_merge_write_throughput": "80mb"
            }
        },
        "retention": {
            "period": "10 years",
            "schedule": "yearly"
        }
    }
    response = requests.post(
        f"{QUICKWIT_ENDPOINT}/api/v1/indexes",
        headers={"Content-Type": "application/json"},
        json=request_data
    )
    print_response(response)


def clear_index(index_name: str) -> None:
    r = requests.put(f"{QUICKWIT_ENDPOINT}/api/v1/indexes/{index_name}/clear")
    print_response(r)


def ingest_into_index(index_name: str, data: List[dict]) -> None:
    data_to_post = '\n'.join(
        [json.dumps(record, sort_keys=True, default=str) for record in data])
    r = requests.post(
        f"{QUICKWIT_ENDPOINT}/api/v1/{index_name}/ingest",
        headers={"Content-Type": "application/x-ndjson"},
        data=data_to_post
    )
    print_response(r)


def get_connection():
    return psycopg.connect(db_string)

def resolve_file_schema_for_docket_ingest(records: List[dict[str, any]]) -> List[dict[str, any]]:
        data = []
        for record in records:
            record['text'] = record.pop('english_text')
            record['source_id'] = str(record.pop('id'))
            record['metadata'] = json.loads(record.pop('mdata'))
            record['timestamp'] = time.time()
            data.append(record)
        return data

def preprocess_id_list_for_sql(ids: List[str]) -> str:
    if len(ids) == 1:
        return f"({ids[0]})"        
    else:
        outstr = " OR ".join(ids)
        return outstr
    
def ingest_files_to_dockets_index(file_ids: Union[str, List[str]]):
    # TODO: valdiate if the file has already been indexed
    conn = get_connection()
    with conn.cursor(row_factory=psycopg.rows.dict_row) as cur:
        ids = preprocess_id_list_for_sql(file_ids)
        cur.execute(f"SELECT * FROM file WHERE ID IN {ids}")
        records = cur.fetchone()

def reindex_whole_db(batch_size: int = 10):
    logger.info("reindexing docokets in quickwit")
    clear_index()
    conn = get_connection()
    with conn.cursor(row_factory=psycopg.rows.dict_row) as cur:
        cur.execute("SELECT * FROM file;")
        while True:
            records = cur.fetchmany(batch_size)
            if not records:
                break
            data_to_insert = resolve_file_schema_for_docket_ingest(records)
            ingest_into_index(data_to_insert)

    logger.info("quickwit indexing request complete")

File: backend-go/quickwit/quickwit.go
package quickwit

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"kessler/objects/files"
	"kessler/objects/timestamp"
	"net/http"
	"os"
	"strings"
	"time"

	"github.com/charmbracelet/log"
)

var quickwitEndpoint = os.Getenv("QUICKWIT_ENDPOINT")

func printResponse(resp *http.Response) {
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		log.Printf("Error reading response body: %v", err)
		return
	}
	log.Printf("quickwit_api_call:\nstatus: %d\nresponse:\n%s", resp.StatusCode, string(body))
}

func CreateIndex(index QuickwitIndex) error {
	jsonData, err := json.Marshal(index)
	if err != nil {
		return fmt.Errorf("error marshaling request data: %v", err)
	}

	resp, err := http.Post(
		fmt.Sprintf("%s/api/v1/indexes", quickwitEndpoint),
		"application/json",
		bytes.NewBuffer(jsonData),
	)
	if err != nil {
		return fmt.Errorf("error making request: %v", err)
	}
	defer resp.Body.Close()

	printResponse(resp)
	return nil
}

func CreateQuickwitProceedingIndex(indexName string) error {
	requestData := QuickwitIndex{
		Version: "0.7",
		IndexID: indexName,
		DocMapping: DocMapping{
			Mode: "dynamic",
			DynamicMapping: DynamicMapping{
				Indexed:    true,
				Stored:     true,
				Tokenizer:  "default",
				Record:     "basic",
				ExpandDots: true,
				Fast:       true,
			},
			FieldMappings: []FieldMapping{
				{Name: "name", Type: "text", Fast: true},
				{Name: "caseNumber", Type: "text", Fast: true},
				{Name: "description", Type: "text", Fast: true},
				{Name: "uuid", Type: "text", Fast: true},
			},
		},
		SearchSettings: SearchSettings{
			DefaultSearchFields: []string{"name"},
		},
		IndexingSettings: IndexingSettings{
			MergePolicy: MergePolicy{
				Type:           "limit_merge",
				MaxMergeOps:    3,
				MergeFactor:    10,
				MaxMergeFactor: 12,
			},
			Resources: Resources{
				MaxMergeWriteThroughput: "80mb",
			},
		},
		Retention: Retention{
			Period:   "10 years",
			Schedule: "yearly",
		},
	}
	err := CreateIndex(requestData)
	return err
}

func ClearIndex(indexName string, nuke bool) error {
	var req *http.Request
	var err error
	if nuke {
		req, err = http.NewRequest(http.MethodDelete, fmt.Sprintf("%s/api/v1/indexes/%s", quickwitEndpoint, indexName), nil)
		if err != nil {
			return fmt.Errorf("error creating request: %v", err)
		}
	} else {
		req, err = http.NewRequest(http.MethodPut, fmt.Sprintf("%s/api/v1/indexes/%s/clear", quickwitEndpoint, indexName), nil)
		if err != nil {
			return fmt.Errorf("error creating request: %v", err)
		}
	}

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return fmt.Errorf("error making request: %v", err)
	}
	defer resp.Body.Close()

	printResponse(resp)
	return nil
}

func ResolveFileSchemaForDocketIngest(complete_files []files.CompleteFileSchema) ([]QuickwitFileUploadData, error) {
	createEnrichedMetadata := func(input_file files.CompleteFileSchema) map[string]interface{} {
		metadata := input_file.Mdata
		metadata["source_id"] = input_file.ID
		metadata["source"] = "ny-puc-energyefficiency-filedocs"
		metadata["conversation_uuid"] = input_file.Conversation.ID.String()
		author_uuids := make([]string, len(input_file.Authors))
		if len(input_file.Authors) == 0 {
			log.Info(fmt.Sprintf("No authors found in file: %v\n", input_file.ID))
		}
		for i, author := range input_file.Authors {
			author_uuids[i] = author.AuthorID.String()
		}
		metadata["author_uuids"] = author_uuids
		// FIXME: IMPLEMENT A date_published FIELD IN PG AND RENDER THIS BASED ON THAT
		dateStr := metadata["date"].(string)
		parsedDate, err := timestamp.CreateRFC3339FromString(dateStr)
		if err != nil {
			metadata["date_filed"] = parsedDate
		}
		return metadata
	}
	var data []QuickwitFileUploadData
	for _, file := range complete_files {
		newRecord := QuickwitFileUploadData{}
		englishText, err := files.EnglishTextFromCompleteFile(file)
		if err != nil {
			// Do nothing, an error here means to text was found.
		}
		if englishText == "" {
			englishText = ""
		}
		newRecord.Text = englishText
		newRecord.SourceID = file.ID
		newRecord.Verified = file.Verified

		newRecord.Metadata = createEnrichedMetadata(file)
		newRecord.Name = file.Name
		date, err := timestamp.CreateRFC3339FromString(file.Mdata["date"].(string))
		if err == nil {
			newRecord.DateFiled = date
		}

		newRecord.Timestamp = time.Now().Unix()
		data = append(data, newRecord)
	}
	// log.Info(fmt.Sprintf("len(data) is %d, len(complete_files) is %d\n", len(data), len(complete_files)))
	if len(data) != len(complete_files) {
		log.Info(fmt.Sprintf("len(data) is %d, len(complete_files) is %d. THEY ARE NOT EUQAL THIS SHOULD NEVER HAPPEN\n", len(data), len(complete_files)))
		return nil, errors.New("ASSERTION ERROR: len(data) is not equal to len(complete_files), dispite no logical way for that to happen!")
	}

	// log.Printf("reformatted data:\n\n%+v\n\n", data)
	return data, nil
}

func MigrateDocketToNYPUC() error {
	query := "metadata.source:(ny-puc-energyefficiency-filedocs)"
	requestData := map[string]interface{}{
		"query":        query,
		"start_offset": 0,
		"max_hits":     40,
		"limit":        1000,
	}

	jsonData, err := json.Marshal(requestData)
	if err != nil {
		return fmt.Errorf("error marshaling request data: %v", err)
	}

	resp, err := http.Post(
		fmt.Sprintf("%s/api/v1/dockets/search", quickwitEndpoint),
		"application/json",
		bytes.NewBuffer(jsonData),
	)
	if err != nil {
		return fmt.Errorf("error making request: %v", err)
	}
	defer resp.Body.Close()

	printResponse(resp)
	return nil
}

func CreateDeleteTask(indexName string, task DeleteTask) error {
	data, err := json.Marshal(task)
	if err != nil {
		return fmt.Errorf("error marshalling delete task request: %v\n\ntask: %s", err, task)
	}
	resp, err := http.Post(
		fmt.Sprintf("%s/api/v1/%s/delete-tasks", quickwitEndpoint, indexName),
		"application/x-ndjson",
		strings.NewReader(string(data)),
	)
	if err != nil {
		return fmt.Errorf("error submitting delete task request: %v", err)
	}
	defer resp.Body.Close()

	printResponse(resp)
	return nil
}
